Avoid NoMethodError in actionview cache_expiry

composerinteralia opened the pull request:
Fixes #36093

When triggering two parallel requests it is possible for the second
request to call `@watcher.execute_if_updated` before `@watcher` has
been set. This change ensures that if `@watcher` has not been set we
take the code path that sets it.

This solution does mean that we may set `@watcher` in the first request
and then set it again in the second request. We may also call
`execute` an extra time. I think this may be OK since it
would only affect the first couple of requests in development.

kaspth commented on actionview/lib/action_view/cache_expiry.rb:
```diff
@@ -20,7 +20,7 @@ def initialize(watcher:)
     def clear_cache_if_necessary
       watched_dirs = dirs_to_watch
-      if watched_dirs != @watched_dirs
+      if watched_dirs != @watched_dirs || @watcher.nil?
```
Let’s put the nil? check first that seems clearer to me.

kaspth reviewed (approved):
I’ve hit this too, thanks for fixing it!

cc @jhawthorn

composerinteralia committed:
Avoid NoMethodError in actionview cache_expiry

Fixes #36093

When triggering two parallel requests it is possible for the second
request to call `@watcher.execute_if_updated` before `@watcher` has
been set. This change ensures that if `@watcher` has not been set we
take the code path that sets it.

This solution does mean that we may set `@watcher` in the first request
and then set it again in the second request. We may also call
`execute` an extra time. I think this may be OK since it
would only affect the first couple of requests in development.

jhawthorn reviewed (approved):
Looks good to me! Thanks.

It might also make sense to wrap the check in a Mutex instead.

kaspth commented:
@jhawthorn now that you mention `Mutex`, isn't this cache expiry enrolled in the autoload interlock via the `executor.to_prepare` which should be thread safe? So now I'm actually questioning why this doesn't get the thread safe for free.

zarqman commented:
@kaspth For me this is triggering from within `ActionDispatch::Executor` (on a series of static asset loads). I have a limited understanding of Rails internals, so I could be wrong here, but I believe that's somewhat before `executor.to_prepare` which is part of `ActionDispatch::Reloader`?
